Sources/CNIOLinux/liburing_shims.c

```diff
@@ -0,0 +1,125 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+// Support functions for liburing
+
+// Check if this is needed, copied from shim.c to avoid possible problems due to:
+// Xcode's Archive builds with Xcode's Package support struggle with empty .c files
+// (https://bugs.swift.org/browse/SR-12939).
+void CNIOLinux_i_do_nothing_just_working_around_a_darwin_toolchain_bug2(void) {}
+
+#ifdef __linux__
+
+#ifdef SWIFTNIO_LIBURING
+
+#define _GNU_SOURCE
+#include <CNIOLinux.h>
+#include <pthread.h>
+#include <unistd.h>
+#include <sys/syscall.h>
+#include <stdlib.h>
+#include <assert.h>
+#include <signal.h>
+#include <signal.h>
+#include <sys/poll.h>
+
+// Definitons from syscall.c in liburing, we'll just do a syscall that will fail
+// to check for availability of the interface - better than checking specific kernel
+// versions as support can be backported to earlier kernels
+
+#ifdef __alpha__
+# ifndef __NR_io_uring_register
+#  define __NR_io_uring_register    537
+# endif
+#else /* !__alpha__ */
+# ifndef __NR_io_uring_register
+#  define __NR_io_uring_register    427
+# endif
+#endif
+
+int _check_syscall_available() {
+    errno = 0;
+    if (syscall(__NR_io_uring_register, 0, IORING_UNREGISTER_BUFFERS, NULL, 0) && errno == ENOSYS) {
+        return -1;
+    }
+    return 0;
+}
+
+int _check_capabilities() {
+    struct io_uring_probe *probe = io_uring_get_probe();
+    int capabilities_check;
+
+    capabilities_check = io_uring_opcode_supported(probe, IORING_OP_POLL_ADD) &
+                         io_uring_opcode_supported(probe, IORING_OP_POLL_REMOVE);
+
+    io_uring_free_probe(probe);
+    return capabilities_check == 1 ? 0 : -1;
+}
+
+int CNIOLinux_check_compatibility()
+{
+    if (_check_syscall_available() != 0) {
+        return -1;
+    }
+    if (_check_capabilities() != 0) {
+        return -1;
+    }
+    return 0;
+}
+
+// We should look at cpu affinity and preferably bind the SQPOLL thread to a specific core
+// and support cgroup / cpu sets on Linux to isolate it - then we should set IORING_SETUP_SQ_AFF here.
+
+static pthread_mutex_t global_ring_mutex = PTHREAD_MUTEX_INITIALIZER;
+static struct io_uring global_ring; // shared small ring to be able to reuse SQPOLL kernel thread
```

Comment:
let's 0 initialise this explicitly with `= {0};`

Reply:
Yep. Actually, even `= {};` is correct btw (struct members that are not specified in the assignment will be initialised as static variables would be in C, which is zero...)

Reply:
Code removed.